Add Service Worker static routing API (#34555)
mdn/content: files/en-us/web/api/installevent/addroutes/index.md

---

> title: "InstallEvent: addRoutes() method"
> short-title: addRoutes()
> slug: Web/API/InstallEvent/addRoutes
> page-type: web-api-instance-method
> browser-compat: api.InstallEvent.addRoutes

Reviewer: ```suggestion
title: "InstallEvent: addRoutes() method"
short-title: addRoutes()
slug: Web/API/InstallEvent/addRoutes
page-type: web-api-instance-method
status:
  - experimental
browser-compat: api.InstallEvent.addRoutes
```

---

> {{APIRef("Navigation API")}}

Reviewer: ```suggestion
{{APIRef("Service Workers API")}}{{SeeCompatTable}}
```

## Syntax

```js-nolint
addRoutes(routerRules)
```

### Parameters

- `routerRules`

  - : A single object, or an array of one or more objects, representing rules for how certain resources should be fetched. Each `routerRules` object contains the following properties:

    - `condition`

      - : An object defining one or more conditions that specify which resources should match this rule. The following properties can be included; if multiple properties are used, a resource must meet all specified conditions to match the rule.
        - `not` {{optional_inline}}
          - : A `condition` object defining conditions that must explicitly **not** be met to match the rule. Conditions defined inside a `not` condition are mutually exclusive with other conditions.
        - `or` {{optional_inline}}
          - : An array of `condition` objects. One set of these defined conditions must be met to match the rule. Conditions defined inside an `or` condition are mutually exclusive with other conditions.
        - `requestMethod` {{optional_inline}}

>           - : A string representing the [HTTP method](/en-US/docs/Web/HTTP/Methods) a request should be sent by for it to match the rule, such as `"GET"`, `"PUT"`, or `"HEAD"`.

Reviewer 2: Your example below uses lower cased name - say `get`. They should match, or somewhere we could say "case insensitive". Ditto for the rest.

Author: done

        - `"network"`
          - : Resources will be loaded from the network.
        - `"race-network-and-fetch-handler"`
          - : Attempts are made to load resources from the network and the service worker's {{DOMxRef("ServiceWorkerGlobalScope.fetch_event", "fetch")}} event handler simultaneously. Whichever one completes first is used.

        The `source` value can also be set to an object containing a single property, `cacheName`, the value of which is a string representing the name of a browser {{domxref("Cache")}}. Matching resources will be loaded from this specific named cache if it exists.

### Return value

None (`undefined`).

### Exceptions

- `TypeError` {{domxref("DOMException")}}
  - : Thrown if one or more of the rules objects inside `routerRules` is invalid, or has a `source` value of `"fetch-event"` when the associated service worker does not have a {{DOMxRef("ServiceWorkerGlobalScope.fetch_event", "fetch")}} event handler.

## Examples

### Route specific requests to the network if the service worker is not running

In the following example, URLs that start with `/articles` are routed to the network if the service worker is not currently running:

> ```js
> addEventListener("install", (event) => {
>   event.addRoutes({
>     condition: {
>       urlPattern: "/articles/*",
>       runningStatus: "not-running",
>     },
>     source: "network",
>   });
> });
> ```

Reviewer 2: Again, why not both cases in one, since it is what you would actually do. I am assuming that would look like this (and worth showing the case with multiple routes):

```suggestion
```js
addEventListener("install", (event) => {
  event.addRoutes([{
    condition: {
      urlPattern: "/articles/*",
      runningStatus: "not-running",
    },
    source: "network",
  },
  {
    condition: {
      urlPattern: "/articles/*",
      runningStatus: "running",
    },
    source: "cache",
  }
  
  ]);
});
```
```

Reviewer 2: Actually @chrisdavidmills I withdraw this suggestion. I understand why you have non-running checks, but not why "running" checks are even defined - in this case your service worker can and probably should handle the case.

Either way, just having the not-running case does the right thing here

---
title: "InstallEvent: addRoutes() method"
short-title: addRoutes()
slug: Web/API/InstallEvent/addRoutes
page-type: web-api-instance-method
status:
  - experimental
browser-compat: api.InstallEvent.addRoutes
---

{{APIRef("Service Workers API")}}{{SeeCompatTable}}

> The **`addRoutes()`** method of the
> {{domxref("InstallEvent")}} interface specifies one or more routes, which define rules for how certain resources should be fetched and where from. This allows you to, for example, bypass a service worker in cases where you always want to fetch a resource from the network or a browser {{domxref("Cache")}}. This is useful when you want to avoid the performance overhead of unnecessary service worker cycles.

Reviewer 2: ```suggestion
The **`addRoutes()`** method of the {{domxref("InstallEvent")}} interface specifies one or more routes, which define rules for fetching specified resources that will be used even prior to service worker startup. This allows you to, for example, bypass a service worker in cases where you always want to fetch a resource from the network or a browser {{domxref("Cache")}}, and avoids the performance overhead of unnecessary service worker cycles.
```

## Syntax

```js-nolint
addRoutes(routerRules)
```

### Parameters

- `routerRules`

>   - : A single object, or an array of one or more objects representing rules for how certain resources should be fetched and where from. Each `routerRules` object contains the following two properties:

Reviewer 2: I think we can drop a few " and where from" for readability.
```suggestion
  - : A single object, or an array of one or more objects representing rules for how certain resources should be fetched. Each `routerRules` object contains the following two properties:
```

    - `condition`

      - : An object defining one or more conditions that specify which resources should match this rule. The following properties can be included; if multiple properties are used, a resource must meet all specified conditions to match the rule.
        - `not` {{optional_inline}}
          - : A `condition` object defining conditions that must explicitly **not** be met to match the rule. Conditions defined inside a `not` condition are mutually exclusive with other conditions.
        - `or` {{optional_inline}}
          - : An array of `condition` objects. One set of these defined conditions must be met to match the rule. Conditions defined inside an `or` condition are mutually exclusive with other conditions.
        - `requestMethod` {{optional_inline}}

>           - : A string representing the [method](/en-US/docs/Web/HTTP/Methods) a request should be sent by for it to match the rule.

Reviewer 2: The implication is that the string will be one of `GET`, `PUT` but perhaps worth providing a few examples to remove all doubt.

```suggestion
          - : A string representing the [HTTP method](/en-US/docs/Web/HTTP/Methods) a request should be sent by for it to match the rule, such as `"GET"`, `"PUT"`, `"HEAD"`. 
```

Author: done, for each one that needs it.

        - `requestMode` {{optional_inline}}
          - : A string representing the [mode](/en-US/docs/Web/API/Request/mode) a request should have for it to match the rule.
        - `requestDestination` {{optional_inline}}
          - : A string representing the [destination](/en-US/docs/Web/API/Request/destination) of a request, i.e. what content type should be requested, for it to match the rule.
        - `runningStatus` {{optional_inline}}
          - : An enumerated value representing the required running status of the service worker for a request to match the rule. Values can be `"running"` or `"not-running"`.
        - `urlPattern` {{optional_inline}}
          - : A {{domxref("URLPattern")}} instance, or a `URLPattern()` constructor [`input`](/en-US/docs/Web/API/URLPattern/URLPattern#input) pattern representing the URLs that match the rule.

    - `source`

>       - : An enumerated value or an object specifying how matching resources will be loaded. Possible enumerated values are:

Reviewer 2: Where, not how.
```suggestion
      - : An enumerated value or an object specifying the source from which matching resources will be loaded. Possible enumerated values are:
```

        - `"cache"`

>           - : Matching resources will be loaded from a browser {{domxref("Cache")}}.

Reviewer 2: In all cases, matching is probably redundant. Consider

```suggestion
          - : Resources will be loaded from a browser {{domxref("Cache")}}.
```

Author: Agreed. I've updated them all to remove the redundancy.

        - `"fetch-event"`

>           - : Matching resources will be loaded via the service worker's {{DOMxRef("ServiceWorkerGlobalScope.fetch_event", "fetch")}} event handler.

Reviewer 2: It might be worth actually noting that this is used with the `running` condition to allow conditions where resources are loaded using the service worker if the worker is running, and otherwise to use a static route fallback.

Author: I have added a sentence to cover this: "This can be combined with the `"runningStatus"` condition to load resources from a service worker if it is running and fall back to a static route on the network if it is not."

        - `"network"`
          - : Matching resources will be loaded from the network.
        - `"race-network-and-fetch-handler"`
          - : Attempts are made to load matching resources from the network and the service worker's {{DOMxRef("ServiceWorkerGlobalScope.fetch_event", "fetch")}} event handler simultaneously. Whichever one completes first is used.

        The value of `source` can also be set to an object containing a single property, `cacheName`, the value of which is a string representing the name of a browser {{domxref("Cache")}}. Matching resources will be loaded from this specific named cache if it exists.

### Return value

None (`undefined`).

### Exceptions

- `TypeError` {{domxref("DOMException")}}
  - : Thrown if one or more of the rules objects inside `routerRules` is invalid, or has a `source` value of `"fetch-event"` when the associated service worker does not have a {{DOMxRef("ServiceWorkerGlobalScope.fetch_event", "fetch")}} event handler.

## Examples

### Route specific requests to a running service worker

> In the following example, URLs that start with `/articles` are routed to the service worker if it is currently running:

Reviewer 2: If you don't add this condition, then the request will force the service worker to start anyway, so "as is" this is a little pointless/raises questions.

You'd use a feature like this if you knew that you could provide a better experience before service worker startup by using a route, but you have a better alternative if it is already started. We need to start from a real example of when you would want to do this and then add both conditions.

Author: OK, so I've turned this on its head to use the network if the service worker is not running:

---

In the following example, URLs that start with `/articles` are routed to the network if the service worker is not currently running:

```js
addEventListener("install", (event) => {
  event.addRoutes({
    condition: {
      urlPattern: "/articles/*",
      runningStatus: "not-running",
    },
    source: "network"
  });
});
```

---

Thinking about it, I'm not sure that you would ever need to add the opposite condition (running, fetch-event), because as you say, it goes to the SW anyway if the route is not added.

Maybe if the service worker is running, we could use the `race-network-and-fetch-handler` value to try to get the fastest result possible out of SW and network if it is running?

```js
addEventListener("install", (event) => {
  event.addRoutes({
    condition: {
      urlPattern: "/articles/*",
      runningStatus: "running",
    },
    source: "fetch-event",
  });
});
```

### Route form post requests to the network

In the following example, [`POST`](/en-US/docs/Web/HTTP/Methods/POST) requests to a form are sent directly to the network and bypass the service worker:

```js
addEventListener("install", (event) => {
  event.addRoutes({
    condition: {
      urlPattern: "/form/*",
      requestMethod: "post",
    },
    source: "network",
  });
});
```

### Route certain image type requests to a named cache

> In the following example, the browser {{domxref("Cache")}} named `"pictures"` is used for fetching files with extensions of `.png` or `.jpg`.

Reviewer 2: 1. Might be worth making that or condition also use GET. That way you can perhaps describe the AND relation of the or and the additional condition.
2. More important, what happens if a cache hit fails?

Author: 1. I'm not really sure what you mean by that. Would you be able to attempt to write an example?
2. I don't know. I suspect it just fails, and it is up to the developer to write their code defensively to deal with that. When the service worker loads, you can tell it to cache any images that aren't already in the cache, and then display the image in the app. The next time they load the app, it'll be cached, so the static route will work.

Reviewer 2: 1. Something like this. Mostly because this is a good example that both the "or" and the other condition must be true.

```js
addEventListener("install", (event) => {
  event.addRoutes({
    condition: {
      or: [{ urlPattern: "*.png" }, { urlPattern: "*.jpg" }],
      requestMethod: "get",
    },
    source: {
      cacheName: "pictures",
    },
  });
});
```

Reviewer 2: 2. I think you need to test this. If it fails and the browser then loads the service worker then your response is fine, but the docs indicate that the existence of aa static route will mean the service worker does not start. There is no defensive coding that the user can write to handle that "first time round".

Author: OK, testing done. 

1. Turns out you can't do that — `or` is mutually exclusive. If you want to do that, you'd have to do it as two separate conditions. I've documented this in my next commit.
2. From my testing, it looks like it automatically falls back to the network if the cache is not available.

Author: I documented that as well.

Reviewer 2: Thank you. That's pretty interesting to know!

### Return value

A {{jsxref("Promise")}} that fulfills with `undefined`.

### Exceptions

- `TypeError` {{domxref("DOMException")}}

>   - : Thrown if one or more of the rules objects inside `routerRules` is invalid, or has a `source` value of `"fetch-event"` when the associated service worker does not have a {{DOMxRef("ServiceWorkerGlobalScope.fetch_event", "fetch")}} event handler. Also thrown if you try to combine `or` with another condition type.

Reviewer 2: According to the spec (as I remember it), every top level must be satisfied for the condition to be true. I.e. it is an AND relationship. You'd use a single OR block to specify a set of conditions that can be true together, and an OR block with other conditions for more complex logic.

It makes sense to me that you might want to have these different kinds of logic, an I don't recall seeing in the spec that combining an OR with other conditions should be an exception. So it may be a bug.

I'm not going to block merging on this, because you've tested, but can you confirm this is the intent and not an accident?

Reviewer 2: Actually, see below, maybe you have https://github.com/mdn/content/pull/34555/files/a76d553d6a7f1da1b222d54e97276661bc95af46..1a4b5688fd9e9b0f4cc2300068b8311bca186220#r1683621231

});
```

> **Note:** If the cache does not exist, the browser defaults to using the network so that the requested resources can still be obtained provided the network is available.

> You can't combine `or` with another condition — this results in a `TypeError`. If for example you wanted to match files with extensions of `.png` or `.jpg` but only when the `requestMethod` is `get`, you'd need to specify two separate conditions:

Reviewer 2: As above, this is great in that it reflects the "truth" of testing, but can we confirm it is the intent of the spec? https://github.com/mdn/content/pull/34555/files/a76d553d6a7f1da1b222d54e97276661bc95af46..1a4b5688fd9e9b0f4cc2300068b8311bca186220#r1683620698
